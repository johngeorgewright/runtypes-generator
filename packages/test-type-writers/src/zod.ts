import { ZodType } from 'zod'

export function zodGuard<T>(validator: ZodType<T>) {
  return (data: unknown): data is T => validator.safeParse(data).success
}
