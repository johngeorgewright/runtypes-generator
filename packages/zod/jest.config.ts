import { InitialOptionsTsJest } from 'ts-jest'

const config: InitialOptionsTsJest = {
  globals: {
    'ts-jest': {
      tsconfig: 'tsconfig.test.json',
    },
  },
  preset: 'ts-jest',
  testEnvironment: 'node',
  setupFiles: ['./zod/jest/globals.ts'],
  snapshotResolver: './zod/jest/snapshotResolver.ts',
  rootDir: '../',
}

export default config
