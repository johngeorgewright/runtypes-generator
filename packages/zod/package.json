{
  "name": "@runtyping/zod",
  "version": "1.3.1",
  "description": "Generate zod from static types & JSON schema.",
  "main": "dist/index.js",
  "scripts": {
    "build": "yarn clean && tsc",
    "clean": "rimraf dist",
    "start": "tsc --watch --preserveWatchOutput",
    "test": "jest"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/johngeorgewright/runtyping/issues"
  },
  "homepage": "https://github.com/johngeorgewright/runtyping#readme",
  "bin": {
    "runtyping": "dist/cli.js"
  },
  "peerDependencies": {
    "typescript": "^4.7.4",
    "zod": "^3.17.9"
  },
  "devDependencies": {
    "@runtyping/test-type-writers": "workspace:*",
    "@semantic-release/commit-analyzer": "9.0.2",
    "@semantic-release/git": "10.0.1",
    "@semantic-release/github": "8.0.7",
    "@semantic-release/npm": "9.0.2",
    "@semantic-release/release-notes-generator": "10.0.3",
    "@types/jest": "29.5.0",
    "@types/json-schema": "7.0.11",
    "@types/node": "18.15.10",
    "jest": "29.5.0",
    "rimraf": "4.4.1",
    "semantic-release": "20.1.3",
    "semantic-release-monorepo": "7.0.5",
    "ts-jest": "29.0.5",
    "ts-node": "10.9.1",
    "typescript": "4.8.4",
    "zod": "3.21.4"
  },
  "dependencies": {
    "@runtyping/generator": "workspace:*",
    "title-case": "3.0.3",
    "ts-morph": "16.0.0",
    "tslib": "2.5.0"
  }
}
