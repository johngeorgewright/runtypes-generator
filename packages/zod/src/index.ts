export { Generator } from '@runtyping/generator'
export * from '@runtyping/generator/dist/runtypes'
