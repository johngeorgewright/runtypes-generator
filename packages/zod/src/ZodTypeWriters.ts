import {
  DeclareType,
  Enum,
  escapeQuottedPropName,
  getGenerics,
  getTypeName,
  Import,
  ImportFromSource,
  propNameRequiresQuotes,
  sortUndefinedFirst,
  Static,
  Tuple,
  TypeWriter,
  TypeWriters,
  Write,
} from '@runtyping/generator'
import { titleCase } from 'title-case'
import { SymbolFlags, ts, Type } from 'ts-morph'
import * as zod from 'zod'

export default class ZodTypeWriters extends TypeWriters {
  #module = 'zod';

  override *defaultStaticImplementation(): TypeWriter {
    yield [Import, { source: this.#module, alias: 'Infer', name: 'infer' }]
    yield [Static, 'Infer<typeof ${name}>']
  }

  override any() {
    return this.#simple('any')
  }

  override *array(type: Type): TypeWriter {
    yield [Import, { source: this.#module, name: 'array' }]
    yield [Write, 'array(']
    yield* this.generateOrReuseType(type.getArrayElementTypeOrThrow())
    yield [Write, ')']
  }

  override boolean() {
    return this.#simple('boolean')
  }

  override *builtInObject(type: Type): TypeWriter {
    yield [
      Import,
      { source: this.#module, name: 'instanceof', alias: 'InstanceOf' },
    ]
    yield [Write, `InstanceOf(${type.getText()})`]
  }

  override *enum(type: Type): TypeWriter {
    const name = getTypeName(type)
    yield [Import, { source: this.#module, name: 'nativeEnum' }]
    yield [ImportFromSource, { name, alias: `_${name}` }]
    yield [Write, `nativeEnum(_${name})`]
  }

  override *enumLiteral(type: Type): TypeWriter {
    const enumTypeName = Enum.getEnumIdentifierNameFromEnumLiteral(type)
    yield [ImportFromSource, { name: enumTypeName, alias: `_${enumTypeName}` }]
    yield* this.#literal(`_${enumTypeName}.${getTypeName(type)}`)
  }

  override *function(): TypeWriter {
    yield [Import, { source: this.#module, alias: 'func', name: 'function' }]
    yield [Write, 'func()']
  }

  override *intersection(type: Type): TypeWriter {
    const [first, ...rest] = type
      .getIntersectionTypes()
      .sort(sortUndefinedFirst)

    if (!first) return yield* this.#simple('undefined')

    yield* this.generateOrReuseType(first)
    for (const item of rest) {
      yield [Write, '.and(']
      yield* this.generateOrReuseType(item)
      yield [Write, ')']
    }
  }

  override *lazy(type: Type): TypeWriter {
    const name = getTypeName(type)
    const alias = `_${name}`
    yield [Import, { source: this.#module, name: 'lazy' }]
    yield [Import, { source: this.#module, name: 'ZodType' }]
    yield [ImportFromSource, { alias, name }]
    yield [DeclareType, `ZodType<${alias}>`]
    yield [Write, 'lazy(() => ']
    yield* this.typeWriter(type)
    yield [Write, ')']
  }

  override literal(type: Type): TypeWriter {
    return this.#literal(type.getText())
  }

  *#literal(value: string): TypeWriter {
    yield [Import, { source: this.#module, name: 'literal' }]
    yield [Write, `literal(${value})`]
  }

  override null() {
    return this.#simple('null')
  }

  override number() {
    return this.#simple('number')
  }

  override *numberIndexedObject(type: Type): TypeWriter {
    yield [Import, { source: this.#module, name: 'record' }]
    yield [Import, { source: this.#module, name: 'number' }]
    yield [Write, 'record(number(), ']
    yield* this.generateOrReuseType(type.getNumberIndexType()!)
    yield [Write, ')']
  }

  override *object(type: Type<ts.ObjectType>): TypeWriter {
    yield [Import, { source: this.#module, name: 'object' }]
    yield [Write, 'object({']

    const typeArguments = getGenerics(type).map((typeArgument) =>
      typeArgument.getText()
    )

    for (const property of type.getProperties()) {
      yield [
        Write,
        `${
          propNameRequiresQuotes(property.getName())
            ? `[\`${escapeQuottedPropName(property.getName())}\`]`
            : property.getName()
        }:`,
      ]
      const propertyType = property.getValueDeclarationOrThrow().getType()
      if (!typeArguments.includes(propertyType.getText()))
        yield* this.generateOrReuseType(propertyType)
      else yield [Write, propertyType.getText()]
      if (property.hasFlags(SymbolFlags.Optional)) yield [Write, '.optional()']
      yield [Write, ',']
    }

    yield [Write, '})']
  }

  override *genericObject(type: Type<ts.ObjectType>): TypeWriter {
    yield [Import, { source: this.#module, alias: 'Infer', name: 'infer' }]
    yield [Import, { source: this.#module, name: 'ZodType' }]
    yield* this.objectFunction(type, 'ZodType', 'Infer')
  }

  override string() {
    return this.#simple('string')
  }

  override *stringIndexedObject(type: Type): TypeWriter {
    yield [Import, { source: this.#module, name: 'record' }]
    yield [Import, { source: this.#module, name: 'string' }]
    yield [Write, 'record(string(), ']
    yield* this.generateOrReuseType(type.getStringIndexType()!)
    yield [Write, ')']
  }

  override *tuple(type: Type): TypeWriter {
    yield [Import, { source: this.#module, name: 'tuple' }]
    yield [Write, 'tuple([']
    for (const element of type.getTupleElements()) {
      yield* this.generateOrReuseType(element)
      yield [Write, ',']
    }
    yield [Write, '])']
  }

  override undefined() {
    return this.#simple('undefined')
  }

  override *union(type: Type): TypeWriter {
    const [first, ...rest] = type.getUnionTypes().sort(sortUndefinedFirst)

    if (!first) return yield* this.#simple('undefined')

    yield* this.generateOrReuseType(first)

    for (const item of rest) {
      yield [Write, '.or(']
      yield* this.generateOrReuseType(item)
      yield [Write, ')']
    }
  }

  override unknown() {
    return this.#simple('unknown')
  }

  override *variadicTuple(type: Type): TypeWriter {
    yield [Import, { source: this.#module, name: 'array' }]

    try {
      const name = getTypeName(type)
      const alias = `_${name}`
      yield [ImportFromSource, { alias, name }]
      yield [Static, alias]
    } catch (error) {
      yield [Static, type.getText()]
    }

    yield [Write, 'array(']
    yield* this.#simple('any')
    yield [
      Write,
      `)
        .min(${Tuple.getTupleMinSize(type)})
        .superRefine((data, ctx) => {`,
    ]
    yield* this.#variadicTupleElements(Tuple.getTupleElements(type))
    yield [
      Write,
      `
        })`,
    ]
  }

  *#variadicTupleElements(types: Tuple.TupleElement[]): TypeWriter {
    let variadicIndex
    for (let i = 0; i < types.length; i++) {
      const { element, variadic } = types[i]
      if (variadic) {
        variadicIndex = i
        yield* this.#variadicTupleVariadicElement(
          element,
          i,
          i === types.length - 1 ? undefined : i - (types.length - 1)
        )
      } else
        yield* this.#variadicTupleElement(
          element,
          variadicIndex === undefined ? i : i - types.length
        )
    }
  }

  *#variadicTupleElement(type: Type, index: number): TypeWriter {
    yield [Import, { source: '@runtyping/zod', name: 'validators' }]
    yield [
      Write,
      `validators.pipeIssues({
        ctx,
        data: ${index >= 0 ? `data[${index}]` : `data.slice(${index})[0]`},
        path: ${index},
        type: `,
    ]
    yield* this.generateOrReuseType(type)
    yield [
      Write,
      `
    });
    `,
    ]
  }

  *#variadicTupleVariadicElement(
    type: Type,
    from: number,
    to?: number
  ): TypeWriter {
    yield [Import, { source: '@runtyping/zod', name: 'validators' }]
    yield [Import, { source: this.#module, name: 'array' }]
    yield [
      Write,
      `validators.pipeIssues({
        ctx,
        data: data.slice(${from}, ${to}),
        path: ${from},
        type: array(`,
    ]
    yield* this.generateOrReuseType(type)
    yield [
      Write,
      `)
    })
    `,
    ]
  }

  override void() {
    return this.#simple('void')
  }

  *#simple(type: keyof typeof zod): TypeWriter {
    if (primitiveNames.includes(type)) {
      const alias = titleCase(type)
      yield [Import, { source: this.#module, name: type, alias }]
      yield [Write, `${alias}()`]
    } else {
      yield [Import, { source: this.#module, name: type }]
      yield [Write, `${type}()`]
    }
  }
}

const primitiveNames = ['any', 'never', 'null', 'undefined', 'unknown', 'void']
