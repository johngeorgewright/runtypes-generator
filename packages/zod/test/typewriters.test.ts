import testTypeWriters from '@runtyping/test-type-writers'
import { z } from 'zod'
import ZodTypeWriters from '../src/ZodTypeWriters'

testTypeWriters<z.ZodTypeAny>({
  createNumberValidator: z.number,
  createStringValidator: z.string,
  createObjectValidator: z.strictObject,
  typeWriters: new ZodTypeWriters(),
  validate(validator, data) {
    validator.parse(data)
  },

  ignore: [
    // TODO: implement advanced version of variadic tuples
    'variadicTuples.B',
    'variadicTuples.C',
    'variadicTuples.D',
  ],
})
