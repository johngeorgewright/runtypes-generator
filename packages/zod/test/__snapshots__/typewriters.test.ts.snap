// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`without static types 1`] = `
"import { literal, number, object, string } from 'zod';

export const One = object({ type: literal(\\"one\\"), value: string(), });
export const Two = object({ type: literal(\\"two\\"), value: number(), });
export const Three = One.or(Two);
"
`;
